sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -176,3 +179,223 @@ case class Crc32(child: Expression) extends UnaryExpression with ImplicitCastInp
     })
   }
 }
+
+/**
+ * A function that calculates hash value for a group of expressions.
+ *
+ * The hash value for an expression depends on its type:
+ *  - null:               0
+ *  - boolean:            0 for true, 1 for false.
```

Comment:
Oh sorry I forgot to update the PR description.
According to the [discussion](https://github.com/apache/spark/pull/10435#discussion_r48482588), we decided to not follow hive for performance concerns. So here I followed the original hash code of internal row so that we don't need to fix a lot of tests.